File: bindings/java/src/test/java/io/omq/UnbindDisconnectTest.java
package io.omq;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.time.Duration;
import org.junit.jupiter.api.Test;

final class UnbindDisconnectTest {
    @Test
    void disconnectThenReconnect() {
        try (Context context = OMQ.context();
             Socket pull = context.socket(SocketType.PULL);
             Socket push = context.socket(SocketType.PUSH)) {
            String endpoint = pull.bind("tcp://127.0.0.1:0");
            push.connect(endpoint);
            push.waitConnected(1, Duration.ofSeconds(5));
            push.disconnect(endpoint);
            push.connect(endpoint);
            push.waitConnected(1, Duration.ofSeconds(5));
            push.send("again");

            assertEquals("again", pull.receive(Duration.ofSeconds(5)).orElseThrow().text());
        }
    }

    @Test
    void unbindThenBindAgain() {
        try (Context context = OMQ.context();
             Socket pull = context.socket(SocketType.PULL);
             Socket push = context.socket(SocketType.PUSH)) {
            String endpoint = pull.bind("tcp://127.0.0.1:0");
            pull.unbind(endpoint);
            pull.bind(endpoint);
            push.connect(endpoint);
            push.waitConnected(1, Duration.ofSeconds(5));
            push.send("after-rebind");

            assertEquals("after-rebind", pull.receive(Duration.ofSeconds(5)).orElseThrow().text());
        }
    }
}

File: bindings/go/unbind_disconnect_test.go
package omq

import (
	"testing"
	"time"
)

func TestDisconnectThenReconnect(t *testing.T) {
	ctx := openTestContext(t)
	defer closeContext(t, ctx)

	pull := newTestSocket(t, ctx, Pull)
	defer closeSocket(t, pull)
	push := newTestSocket(t, ctx, Push)
	defer closeSocket(t, push)

	endpoint, err := pull.Bind("tcp://127.0.0.1:*")
	if err != nil {
		t.Fatal(err)
	}
	if err := push.Connect(endpoint); err != nil {
		t.Fatal(err)
	}
	if _, err := push.WaitConnectedTimeout(1, 5*time.Second); err != nil {
		t.Fatal(err)
	}
	if err := push.Disconnect(endpoint); err != nil {
		t.Fatal(err)
	}
	if err := push.Connect(endpoint); err != nil {
		t.Fatal(err)
	}
	if _, err := push.WaitConnectedTimeout(1, 5*time.Second); err != nil {
		t.Fatal(err)
	}
	if err := push.SendTimeout(String("again"), time.Second); err != nil {
		t.Fatal(err)
	}
	assertRecvString(t, pull, "again")
}

func TestUnbindThenBindAgain(t *testing.T) {
	ctx := openTestContext(t)
	defer closeContext(t, ctx)

	pull := newTestSocket(t, ctx, Pull)
	defer closeSocket(t, pull)
	push := newTestSocket(t, ctx, Push)
	defer closeSocket(t, push)

	endpoint, err := pull.Bind("tcp://127.0.0.1:*")
	if err != nil {
		t.Fatal(err)
	}
	if err := pull.Unbind(endpoint); err != nil {
		t.Fatal(err)
	}
	if _, err := pull.Bind(endpoint); err != nil {
		t.Fatal(err)
	}
	if err := push.Connect(endpoint); err != nil {
		t.Fatal(err)
	}
	if _, err := push.WaitConnectedTimeout(1, 5*time.Second); err != nil {
		t.Fatal(err)
	}
	if err := push.SendTimeout(String("after-rebind"), time.Second); err != nil {
		t.Fatal(err)
	}
	assertRecvString(t, pull, "after-rebind")
}
